src/canvas/static_canvas.class.ts

```diff
@@ -1643,6 +1645,120 @@ export class StaticCanvas<
+   * @param {Number} [options.quality=1] Quality level (0..1). Only used for jpeg.
+   * @param {Number} [options.multiplier=1] Multiplier to scale by, to have consistent
+   * @param {Number} [options.left] Cropping left offset. Introduced in v1.2.14
+   * @param {Number} [options.top] Cropping top offset. Introduced in v1.2.14
+   * @param {Number} [options.width] Cropping width. Introduced in v1.2.14
+   * @param {Number} [options.height] Cropping height. Introduced in v1.2.14
+   * @param {Boolean} [options.enableRetinaScaling] Enable retina scaling for clone image. Introduce in 2.0.0
+   * @param {(object: fabric.Object) => boolean} [options.filter] Function to filter objects.
+   * @return {String} Returns a data: URL containing a representation of the object in the format specified by options.format
+   * @see {@link https://jsfiddle.net/xsjua1rd/ demo}
+   * @example <caption>Generate jpeg dataURL with lower quality</caption>
+   * var dataURL = canvas.toDataURL({
+   *   format: 'jpeg',
+   *   quality: 0.8
+   * });
+   * @example <caption>Generate cropped png dataURL (clipping of canvas)</caption>
+   * var dataURL = canvas.toDataURL({
+   *   format: 'png',
+   *   left: 100,
+   *   top: 100,
+   *   width: 200,
+   *   height: 200
+   * });
+   * @example <caption>Generate double scaled png dataURL</caption>
+   * var dataURL = canvas.toDataURL({
+   *   format: 'png',
+   *   multiplier: 2
+   * });
+   * @example <caption>Generate dataURL with objects that overlap a specified object</caption>
+   * var myObject;
+   * var dataURL = canvas.toDataURL({
+   *   filter: (object) => object.isContainedWithinObject(myObject) || object.intersectsWithObject(myObject)
+   * });
+   */
+  toDataURL(options = {} as TDataUrlOptions): string {
+    const {
+      format = 'png',
+      quality = 1,
+      multiplier = 1,
+      enableRetinaScaling = false,
+    } = options;
+    const finalMultiplier =
+      multiplier * (enableRetinaScaling ? this.getRetinaScaling() : 1);
+
+    return toDataURL(
+      this.toCanvasElement(finalMultiplier, options),
+      format,
+      quality
+    );
+  }
+
+  /**
+   * Create a new HTMLCanvas element painted with the current canvas content.
+   * No need to resize the actual one or repaint it.
+   * Will transfer object ownership to a new canvas, paint it, and set everything back.
+   * This is an intermediary step used to get to a dataUrl but also it is useful to
+   * create quick image copies of a canvas without passing for the dataUrl string
+   * @param {Number} [multiplier] a zoom factor.
+   * @param {Object} [options] Cropping informations
+   * @param {Number} [options.left] Cropping left offset.
+   * @param {Number} [options.top] Cropping top offset.
+   * @param {Number} [options.width] Cropping width.
+   * @param {Number} [options.height] Cropping height.
+   * @param {(object: fabric.Object) => boolean} [options.filter] Function to filter objects.
+   */
+  toCanvasElement(
```

Comment:
why don't we create a new StaticCanvas instance instead of all this mess? I don't believe it will cost in perf and we will benefit in clarity

Reply:
Well actually I tried it but it means that all objects need to be iterated over to set canvas on them so retina will be correct.
So no

Reply:
```ts
  toCanvasElement(
    multiplier = 1,
    { width, height, left, top, filter } = {} as TToCanvasElementOptions
  ): HTMLCanvasElement {
    const scaledWidth = (width || this.width) * multiplier,
      scaledHeight = (height || this.height) * multiplier,
      zoom = this.getZoom(),
      newZoom = zoom * multiplier,
      vp = this.viewportTransform,
      canvasEl = createCanvasElement(),
      objectsToRender = filter ? this._objects.filter(filter) : this._objects;
    canvasEl.width = scaledWidth;
    canvasEl.height = scaledHeight;
    const canvas = new StaticCanvas(canvasEl, {
      width: scaledWidth,
      height: scaledHeight,
      enableRetinaScaling: false,
      viewportTransform: [
        newZoom,
        0,
        0,
        newZoom,
        (vp[4] - (left || 0)) * multiplier,
        (vp[5] - (top || 0)) * multiplier,
      ] as TMat2D,
      backgroundColor: this.backgroundColor,
      backgroundImage: this.backgroundImage,
      backgroundVpt: this.backgroundVpt,
      overlayColor: this.overlayColor,
      overlayImage: this.overlayImage,
      overlayVpt: this.overlayVpt,
      clipPath: this.clipPath,
    });
    canvas.calcViewportBoundaries();
    objectsToRender.forEach((obj) => obj._set('canvas', canvas));
    canvas.renderCanvas(canvasEl.getContext('2d')!, objectsToRender);
    objectsToRender.forEach((obj) => obj._set('canvas', canvas));
    return canvasEl;
  }
```

Reply:
yes this is a long standing issue, there is not easy way to export a render to a different canvas without swapping the canvas values or the object canvas reference. Something that can be solved with a bit of thinking, but also not a large issue as long as this method works